File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC.java
package com.oblador.keychain.cipherStorage;

import android.annotation.TargetApi;
import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyInfo;
import android.security.keystore.KeyProperties;
import androidx.annotation.NonNull;
import android.util.Log;

import com.oblador.keychain.SecurityLevel;
import com.oblador.keychain.exceptions.CryptoFailedException;
import com.oblador.keychain.exceptions.KeyStoreAccessException;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.Charset;
import java.security.InvalidAlgorithmParameterException;
import java.security.Key;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateException;
import java.security.spec.InvalidKeySpecException;
import android.security.keystore.StrongBoxUnavailableException;

import javax.crypto.Cipher;
import javax.crypto.CipherInputStream;
import javax.crypto.CipherOutputStream;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;

@TargetApi(Build.VERSION_CODES.M)
public class CipherStorageKeystoreAESCBC implements CipherStorage {
    public static final String TAG = "KeystoreAESCBC";
    public static final String CIPHER_STORAGE_NAME = "KeystoreAESCBC";
    public static final String DEFAULT_SERVICE = "RN_KEYCHAIN_DEFAULT_ALIAS";
    public static final String KEYSTORE_TYPE = "AndroidKeyStore";
    public static final String ENCRYPTION_ALGORITHM = KeyProperties.KEY_ALGORITHM_AES;
    public static final String ENCRYPTION_BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
    public static final String ENCRYPTION_PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
    public static final String ENCRYPTION_TRANSFORMATION =
            ENCRYPTION_ALGORITHM + "/" +
                    ENCRYPTION_BLOCK_MODE + "/" +
                    ENCRYPTION_PADDING;
    public static final int ENCRYPTION_KEY_SIZE = 256;
    private boolean retry = true;


    @Override
    public String getCipherStorageName() {
        return CIPHER_STORAGE_NAME;
    }

    @Override
    public int getMinSupportedApiLevel() {
        return Build.VERSION_CODES.M;
    }

    @Override
    public SecurityLevel securityLevel() {
        // it can guarantee security levels up to SECURE_HARDWARE/SE/StrongBox
        return SecurityLevel.SECURE_HARDWARE;
    }

    @Override
    public boolean supportsSecureHardware() {
        final String testKeyAlias = "AndroidKeyStore#supportsSecureHardware";

        try {
            SecretKey key = tryGenerateRegularSecurityKey(testKeyAlias);
            return validateKeySecurityLevel(SecurityLevel.SECURE_HARDWARE, key);
        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException | NoSuchProviderException e) {
            return false;
        } finally {
            try {
                removeKey(testKeyAlias);
            } catch (KeyStoreAccessException e) {
                Log.e(TAG, "Unable to remove temp key from keychain", e);
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    public EncryptionResult encrypt(@NonNull String service, @NonNull String username, @NonNull String password, SecurityLevel level, boolean useStrongBox) throws CryptoFailedException {
        service = getDefaultServiceIfEmpty(service);

        try {
            KeyStore keyStore = getKeyStoreAndLoad();

            if (!keyStore.containsAlias(service)) {
                generateKeyAndStoreUnderAlias(service, level, useStrongBox);
            }

            Key key = null;
            try {
                key = keyStore.getKey(service, null);
            } catch (UnrecoverableKeyException ex) {
                ex.printStackTrace();
                // Fix for android.security.KeyStoreException: Invalid key blob
                // more info: https://stackoverflow.com/questions/36488219/android-security-keystoreexception-invalid-key-blob/36846085#36846085
                if (retry) {
                    retry = false;
                    keyStore.deleteEntry(service);
                    return encrypt(service, username, password, level, useStrongBox);
                } else {
                    throw ex;
                }
            }

            byte[] encryptedUsername = encryptString(key, service, username);
            byte[] encryptedPassword = encryptString(key, service, password);

            retry = true;
            return new EncryptionResult(encryptedUsername, encryptedPassword, this);
        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException | NoSuchProviderException | UnrecoverableKeyException e) {
            throw new CryptoFailedException("Could not encrypt data for service " + service, e);
        } catch (KeyStoreException | KeyStoreAccessException e) {
            throw new CryptoFailedException("Could not access Keystore for service " + service, e);
        } catch (Exception e) {
            throw new CryptoFailedException("Unknown error: " + e.getMessage(), e);
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private boolean validateKeySecurityLevel(SecurityLevel level, SecretKey generatedKey) {
        return getSecurityLevel(generatedKey).satisfiesSafetyThreshold(level);
    }

    @TargetApi(Build.VERSION_CODES.M)
    private SecurityLevel getSecurityLevel(SecretKey key) {
        try {
            SecretKeyFactory factory = SecretKeyFactory.getInstance(key.getAlgorithm(), KEYSTORE_TYPE);
            KeyInfo keyInfo;
            keyInfo = (KeyInfo) factory.getKeySpec(key, KeyInfo.class);
            return keyInfo.isInsideSecureHardware() ? SecurityLevel.SECURE_HARDWARE : SecurityLevel.SECURE_SOFTWARE;
        } catch (NoSuchAlgorithmException | NoSuchProviderException | InvalidKeySpecException e) {
            return SecurityLevel.ANY;
        }
    }

    private void generateKeyAndStoreUnderAlias(@NonNull String service, SecurityLevel requiredLevel, boolean useStrongBox) throws NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException, CryptoFailedException {
        // Firstly, try to generate the key as safe as possible (strongbox).
        // see https://developer.android.com/training/articles/keystore#HardwareSecurityModule
        SecretKey secretKey = tryGenerateStrongBoxSecurityKey(service, useStrongBox);
        if (secretKey == null) {
            // If that is not possible, we generate the key in a regular way
            // (it still might be generated in hardware, but not in StrongBox)
            secretKey = tryGenerateRegularSecurityKey(service);
        }

        if (!validateKeySecurityLevel(requiredLevel, secretKey)) {
            try {
                removeKey(service);
            } catch (KeyStoreAccessException e) {
                Log.e(TAG, "Unable to remove key from keychain", e);
            }

            throw new CryptoFailedException("Cannot generate keys with required security guarantees");
        }
    }

    @Override
    public DecryptionResult decrypt(@NonNull String service, @NonNull byte[] username, @NonNull byte[] password) throws CryptoFailedException {
        service = getDefaultServiceIfEmpty(service);

        try {
            KeyStore keyStore = getKeyStoreAndLoad();

            Key key = keyStore.getKey(service, null);
            if (key == null) {
              throw new CryptoFailedException("The provided service/key could not be found in the Keystore");
            }

            String decryptedUsername = decryptBytes(key, username);
            String decryptedPassword = decryptBytes(key, password);

            return new DecryptionResult(decryptedUsername, decryptedPassword, getSecurityLevel((SecretKey) key));
        } catch (KeyStoreException | UnrecoverableKeyException | NoSuchAlgorithmException e) {
            throw new CryptoFailedException("Could not get key from Keystore", e);
        } catch (KeyStoreAccessException e) {
            throw new CryptoFailedException("Could not access Keystore", e);
        } catch (Exception e) {
            throw new CryptoFailedException("Unknown error: " + e.getMessage(), e);
        }
    }

    @Override
    public void removeKey(@NonNull String service) throws KeyStoreAccessException {
        service = getDefaultServiceIfEmpty(service);

        try {
            KeyStore keyStore = getKeyStoreAndLoad();

            if (keyStore.containsAlias(service)) {
                keyStore.deleteEntry(service);
            }
        } catch (KeyStoreException e) {
            throw new KeyStoreAccessException("Failed to access Keystore", e);
        } catch (Exception e) {
            throw new KeyStoreAccessException("Unknown error " + e.getMessage(), e);
        }
    }

    private byte[] encryptString(Key key, String service, String value) throws CryptoFailedException {
        try {
            Cipher cipher = Cipher.getInstance(ENCRYPTION_TRANSFORMATION);
            cipher.init(Cipher.ENCRYPT_MODE, key);
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            // write initialization vector to the beginning of the stream
            byte[] iv = cipher.getIV();
            outputStream.write(iv, 0, iv.length);
            // encrypt the value using a CipherOutputStream
            CipherOutputStream cipherOutputStream = new CipherOutputStream(outputStream, cipher);
            cipherOutputStream.write(value.getBytes("UTF-8"));
            cipherOutputStream.close();
            return outputStream.toByteArray();
        } catch (Exception e) {
            throw new CryptoFailedException("Could not encrypt value for service " + service + ", message: " + e.getMessage(), e);
        }
    }

    private String decryptBytes(Key key, byte[] bytes) throws CryptoFailedException {
        try {
            Cipher cipher = Cipher.getInstance(ENCRYPTION_TRANSFORMATION);
            ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
            // read the initialization vector from the beginning of the stream
            IvParameterSpec ivParams = readIvFromStream(inputStream);
            cipher.init(Cipher.DECRYPT_MODE, key, ivParams);
            // decrypt the bytes using a CipherInputStream
            CipherInputStream cipherInputStream = new CipherInputStream(
                    inputStream, cipher);
            ByteArrayOutputStream output = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            while (true) {
                int n = cipherInputStream.read(buffer, 0, buffer.length);
                if (n <= 0) {
                    break;
                }
                output.write(buffer, 0, n);
            }
            return new String(output.toByteArray(), Charset.forName("UTF-8"));
        } catch (Exception e) {
            throw new CryptoFailedException("Could not decrypt bytes: " + e.getMessage(), e);
        }
    }

    private IvParameterSpec readIvFromStream(ByteArrayInputStream inputStream) {
        byte[] iv = new byte[16];
        inputStream.read(iv, 0, iv.length);
        return new IvParameterSpec(iv);
    }

    private KeyStore getKeyStoreAndLoad() throws KeyStoreException, KeyStoreAccessException {
        try {
            KeyStore keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
            keyStore.load(null);
            return keyStore;
        } catch (NoSuchAlgorithmException | CertificateException | IOException e) {
            throw new KeyStoreAccessException("Could not access Keystore", e);
        }
    }

    @NonNull
    private String getDefaultServiceIfEmpty(@NonNull String service) {
        return service.isEmpty() ? DEFAULT_SERVICE : service;
    }

    @TargetApi(Build.VERSION_CODES.P)
    private SecretKey tryGenerateStrongBoxSecurityKey(String service, boolean useStrongBox) throws NoSuchAlgorithmException,
      InvalidAlgorithmParameterException, NoSuchProviderException {
        // If user set useStrongBox as false skip StrongBox usage
        if (!useStrongBox) {
            return null;
        }
        // StrongBox is only supported on Android P and higher
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
            return null;
        }
        try {
            return generateKey(getKeyGenSpecBuilder(service).setIsStrongBoxBacked(true).build());
        } catch (Exception e) {
          if (e instanceof StrongBoxUnavailableException) {
            Log.i(TAG, "StrongBox is unavailable on this device");
          } else {
            Log.e(TAG, "An error occurred when trying to generate a StrongBoxSecurityKey: " + e.getMessage());
          }
          return null;
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private SecretKey tryGenerateRegularSecurityKey(String service) throws NoSuchAlgorithmException,
      InvalidAlgorithmParameterException, NoSuchProviderException {
        return generateKey(getKeyGenSpecBuilder(service).build());
    }

    // returns true if the key was generated successfully
    @TargetApi(Build.VERSION_CODES.M)
    private SecretKey generateKey(KeyGenParameterSpec spec) throws NoSuchProviderException,
      NoSuchAlgorithmException, InvalidAlgorithmParameterException {
        KeyGenerator generator = KeyGenerator.getInstance(ENCRYPTION_ALGORITHM, KEYSTORE_TYPE);
        generator.init(spec);
        return generator.generateKey();
    }

    @TargetApi(Build.VERSION_CODES.M)
    private KeyGenParameterSpec.Builder getKeyGenSpecBuilder(String service) {
        return new KeyGenParameterSpec.Builder(
                service,
                KeyProperties.PURPOSE_DECRYPT | KeyProperties.PURPOSE_ENCRYPT)
            .setBlockModes(ENCRYPTION_BLOCK_MODE)
            .setEncryptionPaddings(ENCRYPTION_PADDING)
            .setRandomizedEncryptionRequired(true)
            //.setUserAuthenticationRequired(true) // Will throw InvalidAlgorithmParameterException if there is no fingerprint enrolled on the device
            .setKeySize(ENCRYPTION_KEY_SIZE);
    }
}

File: android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.kt
package com.oblador.keychain.cipherStorage

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.oblador.keychain.KeychainModule.KnownCiphers
import com.oblador.keychain.SecurityLevel
import com.oblador.keychain.cipherStorage.CipherStorageKeystoreAesCbc.IV.IV_LENGTH
import com.oblador.keychain.resultHandler.ResultHandler
import com.oblador.keychain.exceptions.KeychainException
import java.io.IOException
import java.security.GeneralSecurityException
import java.security.Key
import java.security.spec.KeySpec
import java.util.concurrent.atomic.AtomicInteger
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec

class CipherStorageKeystoreAesCbc(reactContext: ReactApplicationContext) :
    CipherStorageBase(reactContext) {

    // region Constants
    /** AES */
    companion object {
        const val ALGORITHM_AES = KeyProperties.KEY_ALGORITHM_AES

        /** CBC */
        const val BLOCK_MODE_CBC = KeyProperties.BLOCK_MODE_CBC

        /** PKCS7 */
        const val PADDING_PKCS7 = KeyProperties.ENCRYPTION_PADDING_PKCS7

        /** Transformation path. */
        const val ENCRYPTION_TRANSFORMATION = "$ALGORITHM_AES/$BLOCK_MODE_CBC/$PADDING_PKCS7"

        /** Key size. */
        const val ENCRYPTION_KEY_SIZE = 256

        const val DEFAULT_SERVICE = "RN_KEYCHAIN_DEFAULT_ALIAS"
    }

    // endregion

    // region Configuration
    override fun getCipherStorageName(): String = KnownCiphers.AES_CBC

    /** API23 is a requirement. */
    override fun getMinSupportedApiLevel(): Int = Build.VERSION_CODES.M

    /** It can guarantee security levels up to SECURE_HARDWARE/SE/StrongBox */
    override fun securityLevel(): SecurityLevel = SecurityLevel.SECURE_HARDWARE

    /** Biometry is Not Supported. */
    override fun isAuthSupported(): Boolean = false

    /** AES. */

    override fun getEncryptionAlgorithm(): String = ALGORITHM_AES

    /** AES/CBC/PKCS7Padding */

    override fun getEncryptionTransformation(): String = ENCRYPTION_TRANSFORMATION

    /** Override for saving the compatibility with previous version of lib. */
    override fun getDefaultAliasServiceName(): String = DEFAULT_SERVICE

    // endregion

    // region Overrides

    @Throws(KeychainException::class)
    override fun encrypt(
        handler: ResultHandler,
        alias: String,
        username: String,
        password: String,
        level: SecurityLevel
    ) {

        throwIfInsufficientLevel(level)

        val safeAlias = getDefaultAliasIfEmpty(alias, getDefaultAliasServiceName())
        val retries = AtomicInteger(1)

        try {
            val key = extractGeneratedKey(safeAlias, level, retries)

            val result = CipherStorage.EncryptionResult(
                encryptString(key, username), encryptString(key, password), this
            )
            handler.onEncrypt(result, null)
        } catch (fail: Throwable) {
            throw KeychainException("Could not encrypt data with alias: $alias, error: ${fail.message}", fail)
        }
    }


    /** Redirect call to [decrypt] method. */
    @Throws(KeychainException::class)
    override fun decrypt(
        handler: ResultHandler,
        alias: String,
        username: ByteArray,
        password: ByteArray,
        level: SecurityLevel
    ) {

        throwIfInsufficientLevel(level)

        val safeAlias = getDefaultAliasIfEmpty(alias, getDefaultAliasServiceName())
        val retries = AtomicInteger(1)

        try {
            val key = extractGeneratedKey(safeAlias, level, retries)

            val results = CipherStorage.DecryptionResult(
                decryptBytes(key, username), decryptBytes(key, password), getSecurityLevel(key)
            )
            handler.onDecrypt(results, null)
        } catch (e: GeneralSecurityException) {
            throw KeychainException("Could not decrypt data with alias: $alias, error: ${e.message}", e)
        } catch (fail: Throwable) {
            handler.onDecrypt(null, fail)
        }
    }

    // endregion

    // region Implementation

    /** Get encryption algorithm specification builder instance. */

    @Throws(GeneralSecurityException::class)
    override fun getKeyGenSpecBuilder(
        alias: String
    ): KeyGenParameterSpec.Builder {
        val purposes = KeyProperties.PURPOSE_DECRYPT or KeyProperties.PURPOSE_ENCRYPT

        return KeyGenParameterSpec.Builder(alias, purposes)
            .setBlockModes(BLOCK_MODE_CBC)
            .setEncryptionPaddings(PADDING_PKCS7)
            .setRandomizedEncryptionRequired(true)
            .setKeySize(ENCRYPTION_KEY_SIZE)
    }

    /** Get information about provided key. */

    @Throws(GeneralSecurityException::class)
    override fun getKeyInfo(key: Key): KeyInfo {
        val factory = SecretKeyFactory.getInstance(key.algorithm, KEYSTORE_TYPE)
        val keySpec: KeySpec = factory.getKeySpec(key as SecretKey, KeyInfo::class.java)

        return keySpec as KeyInfo
    }

    /** Try to generate key from provided specification. */

    @Throws(GeneralSecurityException::class)
    override fun generateKey(spec: KeyGenParameterSpec): Key {
        val generator = KeyGenerator.getInstance(getEncryptionAlgorithm(), KEYSTORE_TYPE)

        // initialize key generator
        generator.init(spec)

        return generator.generateKey()
    }

    /** Decrypt provided bytes to a string. */

    @Throws(GeneralSecurityException::class, IOException::class)
    override fun decryptBytes(
        key: Key,
        bytes: ByteArray,
        handler: DecryptBytesHandler?
    ): String {
        val cipher = getCipher()

        return try {
            // read the initialization vector from bytes array
            val iv = ByteArray(IV_LENGTH)

            if (IV_LENGTH >= bytes.size)
                throw IOException("Insufficient length of input data for IV extracting.")

            System.arraycopy(bytes, 0, iv, 0, IV_LENGTH)

            val spec = IvParameterSpec(iv)
            cipher.init(Cipher.DECRYPT_MODE, key, spec)

            // Decrypt the bytes using cipher.doFinal()
            val decryptedBytes = cipher.doFinal(bytes, IV.IV_LENGTH, bytes.size - IV.IV_LENGTH)
            String(decryptedBytes, UTF8)
        } catch (fail: Throwable) {
            Log.w(LOG_TAG, fail.message, fail)
            throw fail
        }
    }

    // endregion

    // region Initialization Vector encrypt/decrypt support

    /** Initialization vector support. */
    object IV {
        /** Encryption/Decryption initialization vector length. */
        const val IV_LENGTH = 16

        /** Save Initialization vector to output stream. */
        val encrypt = EncryptStringHandler { cipher, key, output ->
            cipher.init(Cipher.ENCRYPT_MODE, key)
            val iv = cipher.iv
            output.write(iv, 0, iv.size)
        }

        /** Read initialization vector from input stream and configure cipher by it. */
        val decrypt = DecryptBytesHandler { cipher, key, input ->
            val iv = ByteArray(IV_LENGTH)
            val result = input.read(iv, 0, IV_LENGTH)

            if (result != IV_LENGTH) throw IOException("Input stream has insufficient data.")

            val spec = IvParameterSpec(iv)

            cipher.init(Cipher.DECRYPT_MODE, key, spec)
        }

    }


    @Throws(GeneralSecurityException::class, IOException::class)
    override fun encryptString(key: Key, value: String): ByteArray =
        encryptString(key, value, IV.encrypt)


    @Throws(GeneralSecurityException::class, IOException::class)
    override fun decryptBytes(key: Key, bytes: ByteArray): String =
        decryptBytes(key, bytes, IV.decrypt)

    // endregion
}
